Enable ildasm/ilasm round trip testing

kyulee1 opened the pull request:
This enables il{d}asm round-trip testing for Windows.
In order to test il{d}asm on xPlatforms, I injected il{d}asm running on the target binary before running it.
This means this step is a part of the runtest not the buildtest -- unlike runtest, currently buildtest is only on Windows.
1. Test batch file is extended to inject il{d}asm running
2. RunTest has now an optional flag "noasmgen" to fall back the original run without the round-tripping.
3. Test timeout is increased since the round-tripping is part of runtime.
4. Did some clean-up on CLRTest.Execute.Batch.targets
5. Bail out the round-tripping for the tests that are built against mscorlib instead of .Net Core.
6. Added /raweh to ildasm due to https://github.com/dotnet/coreclr/issues/2481

Most tests (>5K) are covered with this change.
The total test time is increased from 2287 to 2438 on debug build. Expecting the gap is smaller on release build/run.

jkotas commented:
I think it is overkill to do ilasm/ildasm roundtrip testing on all tests all the time.

jkotas commented:
Also, the unconditional roundtripping is introducing a test hole: The ildam/ilasm roundtrip will not produce binaries that are byte-to-byte identical to the original so we will miss low-level patterns that are produced by the C# compiler, but not by ilasm/ildasm.

The ilasm/ildasm roundtrip testing should be off by default. As a smoke test, we can do it on a small number of tests (~10) in regular test runs.

kyulee1 commented:
We could run both the il{d}asm roundtrip binary and the original binary to avoid a test hole.
I agree the roundtrip is off by default and enable only subset in the regular test run.
But I'd like to enable all as possible in a daily or weekly stress test.
@mmitche Do we have nightly/weekly run?

mmitche commented:
No, though we can create one.

Sent from my Windows Phone
________________________________
From: Kyungwoo Lee<mailto:notifications@github.com>
Sent: ‎12/‎30/‎2015 6:51 AM
To: dotnet/coreclr<mailto:coreclr@noreply.github.com>
Cc: Matt Mitchell<mailto:mmitche@microsoft.com>
Subject: Re: [coreclr] Enable ildasm/ilasm round trip testing (#2487)


We could run both the il{d}asm roundtrip binary and the original binary to avoid a test hole.
I agree the roundtrip is off by default and enable only subset in the regular test run.
But I'd like to enable all as possible in a daily or weekly stress test.
@mmitche<https://github.com/mmitche> Do we have nightly/weekly run?

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/coreclr/pull/2487#issuecomment-168014093>.

kyulee1 commented:
@jkotas I swithced it off by default. PTAL

jkotas commented:
Could you please rename `asmgen` to something like `ilasmroundtrip` so that it is more obvious what it does?

jkotas commented on tests/src/CLRTest.Execute.Batch.targets:
```diff
@@ -46,6 +46,35 @@ if not exist "$(MSBuildProjectName).ni.exe" "%CORE_ROOT%\crossgen.exe" /Platform
     </PropertyGroup>
   </Target>
```
Is a matching change needed in CLRText.Execute.Bash.targets for Unix?

kyulee1 replied:
This is for Windows only and will follow UNIX change in the next check in after testing.

Sent from Outlook<http://aka.ms/Ox5hz3>




On Wed, Dec 30, 2015 at 6:03 PM -0800, "Jan Kotas" <notifications@github.com<mailto:notifications@github.com>> wrote:


In tests/src/CLRTest.Execute.Batch.targets<https://github.com/dotnet/coreclr/pull/2487#discussion_r48643999>:

> @@ -46,6 +46,35 @@ if not exist "$(MSBuildProjectName).ni.exe" "%CORE_ROOT%\crossgen.exe" /Platform
>      </PropertyGroup>
>    </Target>


Is a matching change needed in CLRText.Execute.Bash.targets for Unix?

-
Reply to this email directly or view it on GitHub<https://github.com/dotnet/coreclr/pull/2487/files#r48643999>.

jkotas commented on tests/runtest.proj:
```diff
@@ -264,7 +268,7 @@ namespace $([System.String]::Copy($(Category)).Replace(".","_").Replace("\","").
                   System.IO.Directory.CreateDirectory(_Global.reportBase + testSubfolder)%3B
 
-                  ret = wrapper.RunTest(testExecutable, outputFile, errorFile)%3B
+                  ret = wrapper.RunTest(testExecutable, outputFile, errorFile, "$(TestArguments)")%3B
```
I do not think that the tests should have arguments. The plan that we have been so far on was that the test are self-contained units.

There is prior art to what you are doing, e.g. with crossgen or with compiling tests as /debug- vs. /debug+. We have been adding these special execution modes as either:
- Global properties: set a global variable and run all tests in special execution mode
- Test local properties: special execution mode is specified in the test project file

I expect that we will be adding more of these special execution modes in future. If we keep driving them using global lists on the side, it will become extremely hard to reason about what the test does - one would need to look at the test project file, and also at a bunch of other global lists to figure out what special execution modes are turned on for it. It is not a good place to be in. We need to be in a place where you can tell what the test does by just looking at its project file.

jkotas commented:
LGTM otherwise

kyulee1 commented:
@jkotas From your feedbacks,
Renamed 'asmgen' to 'ilasmroundtrip' in all instances.
Removed argument to test script file/runtest. Instead 'ilasmroundtrip' optional build flag injects this additional part to the script globally.

jkotas commented on tests/ilasmroundtrip.targets:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+    <ItemGroup>
+        <IlasmRoundTripList Include="$(BaseOutputPathWithConfig)\JIT\BBT\Scenario4\Not-Int32\" />
```
I am not sure about this global list. Per my earlier comment - I think it would be better to set the ilasmroundtrip property in each test project file, so that it is obvious from the test project file what the test is actually doing.

jkotas commented:
LGTM

kyulee1 commented:
@jkotas I deleted the global ilasmroundtrip.targets and add the property to each project file instead.

jkotas commented:
:+1:
